autokeras/hypermodel/processor.py

```diff
@@ -1,33 +1,94 @@
 import tensorflow as tf
 import numpy as np
+from sklearn.feature_extraction import text
+from sklearn import feature_selection
+from tensorflow.python.util import nest
 
-from autokeras import utils
+from autokeras import const
 from autokeras.hypermodel import hyper_block as hb_module
 
 
 class HyperPreprocessor(hb_module.HyperBlock):
+    """Hyper preprocessing block base class."""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self._hp = None
 
     def build(self, hp, inputs=None):
+        """Build into part of a Keras Model.
+
+        Since they are for preprocess data before feeding into the Keras Model,
+        they are not part of the Keras Model. They only pass the inputs
+        directly to outputs.
+        """
         return inputs
 
-    def fit_transform(self, hp, inputs):
-        self.fit(hp, inputs)
-        return self.transform(hp, inputs)
+    def set_hp(self, hp):
+        """Set Hyperparameters for the Preprocessor.
+
+        Since the `update` and `transform` function are all for single training
+        instances instead of the entire dataset, the Hyperparameters needs to be
+        set in advance of call them.
 
-    def fit(self, hp, inputs):
+        Args:
+            hp: Hyperparameters. The hyperparameters for tuning the preprocessor.
+        """
+        self._hp = hp
+
+    def update(self, x):
+        """Incrementally fit the preprocessor with a single training instance.
+
+        Args:
+            x: EagerTensor. A single instance in the training dataset.
+        """
         raise NotImplementedError
 
-    def transform(self, hp, inputs):
+    def transform(self, x):
+        """Incrementally fit the preprocessor with a single training instance.
+
+        Args:
+            x: EagerTensor. A single instance in the training dataset.
+
+        Returns:
+            A transformed instanced which can be converted to a tf.Tensor.
+        """
         raise NotImplementedError
 
+    def output_types(self):
+        """The output types of the transformed data, e.g. tf.int64.
+
+        The output types are required by tf.py_function, which is used for transform
+        the dataset into a new one with a map function.
+
+        Returns:
+            A tuple of data types.
+        """
+        raise NotImplementedError
+
+    def output_shape(self):
+        """The output shape of the transformed data.
+
+        The output shape is needed to build the Keras Model from the AutoModel.
+        The output shape of the preprocessor is the input shape of the Keras Model.
+
+        Returns:
+            A tuple of ints or a TensorShape.
+        """
+        raise NotImplementedError
+
+    def post_fit(self):
```

Comment:
Maybe `finalize` would be clearer. There is not `fit` method in this class. `finalize` and "finalizer" are fairly standard concepts. https://en.wikipedia.org/wiki/Finalizer